Enn

tuanle618 opened the pull request:
Hi @wsad1 ,
as discussed in #2824 , I update the code for the equivariant conv. 
Changes been made:

1. for an (optional) batch, create the fully-connected graph edge_index in COO format. Crucial point is the offset parameter `ptr` which can optionally be passed in forward. As the `EquivariantConv` is a module in a GNN, I included the option to pass the `fc_edge_index` right away, as this can be calculated once, and passed forward for the consecutive `EquivariantConv` message-passing layers.
2. handle edge-attributes in the fully-connected graph. Essentially, I zero-pad the 0-vector for the "fake" edge-indices in the fully-connected graph, that are non-present but still needed for updating the positional features. For the `self.local_nn` then, users should make sure that the `self.local_nn` does not contain a bias vector for the first linear layer (or at least the biases at positional-dim, where the edge_attr are considered should be 0).
3. added tests to make sure the `EquivariantConv` updates the node embeddings `h` based on `edge_index` and the positional features `x` based on `fc_edge_index`

tuanle618 committed:
update equivariant conv to handle fully-connected graph

tuanle618 committed:
small change in function name

wsad1 commented:
@tuanle618 , thanks for all the work, it looks really good. I have a few questions/thoughts.

1. The `get_fully_connected_edges_in_batch`. Could we change it to something like 
```
def get_fully_connected_edges_in_batch (batch: Tensor, edge_index: Tensor,
                                       add_self_loops: bool = False,
                                       fill_value: float = 0.0,
                                       edge_attr: OptTensor = None) -> Tuple[Tensor, Tensor, OptTensor]:

```
So the function takes batch as input, that way the user doesn't need to create batch_num and ptr. Also the function returns 3 tensors `all_edge_index` fully connected edge index, `edge_mask` this tells us the position of the original edge index in `all_edge_index` and `all_edge_attr` edge attributes with the original edge attributes in `all_edge_attr[edge_mask]` and the other locations filled by `fill_value`.  These changes make the function consistent with other functions in `torch_geometric.utils` and it removes the requirement that the original edge indices need to be at the beginning of `all_edge_index`.

2. Another advantage of returning `edge_mask` is that it might simplify `forward` and `aggregate`.
```
    def forward(self, x: OptTensor,
                pos: Tensor,
                fc_edge_index: Tensor, 
                edge_mask: Tensor,
                vel: OptTensor = None,
                edge_attr: OptTensor = None) -> Tuple[Tensor, Tuple[Tensor, OptTensor]]:

    def aggregate(self, inputs: Tuple[Tensor, Tensor],
                  index: Tensor, edge_mask: Tensor) -> Tuple[Tensor, Tensor]:
        node_aggr_messages = scatter(inputs[0][edge_mask], index[edge_mask], 0, reduce=self.aggr)
        node_aggr_positional = scatter(inputs[1], index, 0, reduce="mean")
        return (node_aggr_messages, node_aggr_positional)
```
In the above `forward` call the `fc_edge_index` is not optional, we can mention in the comments that `get_fully_connected_edges_in_batch` could be used to generate it, I think this takes away a bit of flexibility but makes it easier to use/understand . Further `aggregate` can use the `edge_mask` to aggregate the message.

Let me know your thoughts on this, or if you have other ideas.

tuanle618 commented:
Hi @wsad1 ,
sorry for late reply. I am going to continue working on this PR by the end of this week. But essentially, I agree with you. My initial motive for including the `batch_num_nodes` (and `ptr`) argument was just to avoid recomputing these 1-d tensors over again for each `EquivariantConv` layer. But making it more user-friendly and consistent with the PyG API, is definitely better and preferred.

I like the idea with the `edge_mask`. Gonna push some changes soon and fix the lint-problems so the checks are passing.
Thanks again for your feedback.

tuanle618 committed:
update EquivariantConv using edge-mask and fix test CI errors

tuanle618 committed:
let user enable Not to use FC-graph

tuanle618 committed:
fix lint errors

tuanle618 committed:
fix lint error test

tuanle618 commented:
Hi @wsad1 ,

it took me longer to work on this PR again, but I changed the code for the `EquivariantConv` using the `edge_mask`.
A benefit now is that the `self.aggregate` function does not require the `original_index` 1D-tensor, but can be queried from the 1D-`fc_edge_index` tensor using the `edge_mask`.
As the user might want to still rely on the **local neighbourhood** only, when updating the **positional** embeddings, (as in your previous) implementation, I added the `use_fc` argument, when instantiating the `EquivariantConv`. This might be preferred, when working on really large graphs, and computing the message-embeddings might be a bottleneck (computation and memory-wise).
A way to still have a "larger" receptive field when updating the positional features, could be by creating a radius-graph that exceed the local neighbourhood provided by `edge_index` - but this would require that the user inputs `edge_index` and `larger_edge_index` by her/himself.

Let me know if the PR is ok.


Kind regards, and sorry for the delay.

Tuan

wsad1 commented:
Thanks for the great work @tuanle618 . I'll try to review this this weekend or the next.
